### PR TITLE
#162170675 Article Rating Feature

### PR DESCRIPTION
**Pull request by marcdomain**
#### What does this PR do?
Handles implementation of article rating feature

#### Description of Task to be completed?
`POST /article/slug/rating`
- Test the feature using `jest` and `react-testing-library`
- Create component for article rating
- Consume the API endpoint

#### How should this be manually tested?
- Clone this repository `git@github.com:andela/balder-ah-frontend.git`
- cd into the directory on your local machine
- On your command line, run `npm install` to install the project npm packages
- Run `npm run start:dev` to start the app
- On your web browser, visit an article page on `/article/slug` and checkout the implementation by clicking on the star icons to rate the article on a scale of 1 to 5.
- Run `npm test` to test the app
- Run `npm test -- --coverage` to view the test coverage

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#162170675](https://www.pivotaltracker.com/story/show/162170675)

#### Screenshots
screen shot 2019-01-24 at 12 56 58
screen shot 2019-01-24 at 12 58 03

### CODE REVIEW

File: src/actions/rating.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+import authUtils from '../utils/auth';
+
+const postArticleRating = (ratingValue, slug) => async (dispatch) => {
+  try {
+    const userToken = authUtils.getUserToken();
+
+    axios.defaults.headers.common.Authorization = userToken;
+
+    await axios.post(`/articles/${slug}/rating`, ratingValue);
+  } catch (response) {
```
Comment (oluwajuwon) [inline src/actions/rating.js]: @marcdomain this catch block here, are you using the variable name `response` to catch the error? 
Another thing, if the you're saying `const errorMessage = response` on line 13 below, will `errorMessage` be a string or an object?

**Review by oluwajuwon — COMMENTED** [review]
Hi @marcdomain Nice job implementing this feature.
I left a comment on your PR to seek clarity, please check it out

File: src/actions/rating.js (hunk 1 above)
Comment (marcdomain) [reply src/actions/rating.js]: Good observation. 
The catch block is actually meant to catch errors for users that are not logged in and trying to rate the article. For now the catch error is not dispatched (it's redundant) as the `isLogedIn` in our `store` is been fixed in Luqman's PR. So for now, if you're not logged in, you won't get an error message, even though your rating would not be submitted to the API (you only see the frontend animation that gives you a feel that you've rated the article). The dispatch in the catch block would be fixed as Luqman's PR is merged.
This implementation actually has no issues as it is, but would be made more elegant briefly.

File: src/actions/rating.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { ARTICLE_RATING_ERROR } from './types';
+import axios from '../utils/axiosInstance';
+import authUtils from '../utils/auth';
+
+const postArticleRating = (ratingValue, slug) => async (dispatch) => {
+  try {
+    const userToken = authUtils.getUserToken();
```
Comment (frostyblok) [inline src/actions/rating.js]: We have this covered with the interceptors in our utils already

**Review by frostyblok — COMMENTED** [review]
Hi Marcus, good job on delivering on the task. I left a comment on your PR, please check it out to make changes.

**Review by akhilome — APPROVED** [review]

File: src/components/presentation/RatingArticles.test.jsx [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
+import { fireEvent, render } from 'react-testing-library';
+import { Provider } from 'react-redux';
+import 'jest-dom/extend-expect';
```
Comment (Johnsonojo) [inline src/components/presentation/RatingArticles.test.jsx]: Hi @marcdomain, Nice implementation of your feature. Consider removing the import on line 4. There is a setupTest file already handling for that.

**Review by Johnsonojo — CHANGES_REQUESTED** [review]

File: src/actions/rating.js (hunk 1 above)
Comment (oluwajuwon) [reply src/actions/rating.js]: ok then 👍

**Review by oluwajuwon — APPROVED** [review]

**Review by frostyblok — APPROVED** [review]

File: src/components/presentation/RatingArticles.test.jsx (hunk 3 above)
Comment (marcdomain) [reply src/components/presentation/RatingArticles.test.jsx]: Thanks for the heads up. It has been updated

File: src/components/article/ViewArticle.jsx [hunk 4]
```diff
@@ -16,6 +17,12 @@ class ViewArticle extends React.Component {
   componentDidMount = async () => {
     const { match: { params }, getArticle: viewArticle } = this.props;
     await viewArticle(params.slug);
+    const { message, history } = this.props;
+    if (typeof message !== 'string') {
+      setTimeout(() => {
```
Comment (okoroemeka) [inline src/components/article/ViewArticle.jsx:30]: @marcdomain Nice work. But I think making use of setTimeout here will add to the time it going to take for the page to load when the user is directed to it. What do you think?

**Pull request merged**